coqpyt/coq/proof_file.py

```diff
@@ -376,69 +419,57 @@ def __is_proof_term(self, step: Step):
             TermType.CLASS,
             TermType.SCHEME,
             TermType.VARIANT,
+            TermType.OTHER,
         ]
 
+    def __is_end_proof(self, step: Step):
+        return self.context.expr(step)[0] in ["VernacEndProof", "VernacExactProof"]
+
     def __check_proof_step(self, step: Step, goals: Optional[GoalAnswer] = None):
         # Last step of the file
         if step.text.strip() == "":
             return
-        # Forward steps outside of proofs are ignored
-        # NOTE: CoqFile raises an exception for backward steps
-        if not self.in_proof and (goals is None or goals.goals is None):
-            return
 
         # Found [Qed]/[Defined]/[Admitted] or [Proof <exact>.]
-        if self.context.expr(step)[0] in ["VernacEndProof", "VernacExactProof"]:
+        if self.__is_end_proof(step):
             self.__handle_end_proof(step, goals)
-        # Regular proof steps for the last open proof
-        elif self.context.term_type(step) == TermType.OTHER:
-            self.__handle_proof_step(step, goals)
-        # FIXME: An exception is raised because we found a term definition
-        # and CoqFile does not remove the term when stepping backwards
-        elif goals is None:
-            raise NotImplementedError(
-                "Found term definition while going backwards in proof"
-            )
-        # Ignore term definitions that do not generate new proof goals
         elif self.__is_proof_term(step):
             self.__handle_proof_term(step, goals)
+        # Avoids Tactics, Notations, Inductive...
+        elif self.context.term_type(step) == TermType.OTHER:
+            self.__handle_proof_step(step, goals)
 
     def __forward_step(self, goals: GoalAnswer):
         self.__aux_file.append(self.prev_step.text)
         self.__check_program()
-        self.__check_proof_step(self.prev_step, goals)
+        if self.in_proof or self.__is_end_proof(self.prev_step):
+            self.__check_proof_step(self.prev_step, goals)
 
     def __backward_step(self):
         self.__aux_file.truncate(self.curr_step.text)
-        self.__check_proof_step(self.curr_step)
+        if len(self.open_proofs) > 0 or self.__is_end_proof(self.curr_step):
```

Comment:
I think I get it. It's because we still need to remove the proof from the `__open_proofs`, even if the step was already taken (i.e., `in_proof` is false). In that case, I think the length of `__open_proofs` should be checked in `__forward_step`, as well. I think it makes sense and it's a good idea to keep the functions as similar as possible. Do you agree?

Reply:
Actually there were two reasons:
- The first reason is the one you mentioned
- The second reason is because of performance. `in_proof` can take about 0.5 seconds if the goals were not already obtained.

In `__forward_step`, the reason why I didn't use `__open_proofs` is because of the steps that define the terms. When you create a Theorem, you still don't have `__open_proofs`. We could do the reverse of what we do in the backward steps with the `__is_end_proof` and use the `__is_proof_term`, but the `__is_proof_term` does not have guarantees that a proof was generated. For that reason, the only option that I found was to use `in_proof`.